python/mspasspy/db/normalize.py

```diff
@@ -0,0 +1,2052 @@
+                # Hence we copy all.
+                for key in doc:
+                    if self.prepend_collection_name:
+                        newkey = self.collection + "_" + key
+                        d[newkey] = doc[key]
+                    else:
+                        d[key] = doc[key]
+
+            return d
+        else:
+            # land here if d was not a valid datum.
+            raise TypeError("ID_matcher.normalize:  received invalid data type")
+
+
+def _channel_composite_key(net, sta, chan, loc, separator="_"):
+    """
+    Returns a composite key that is unique for a seed channel defined
+    (by default) a net_sta_chan_loc.  Optional separator argument
+    can be used to use a different separator character.
+    Loc is often null so choose to not include a trailing separator
+    when that is the case.  Similarly, when chan is a zero length
+    string it is omitted.  That allows this same function to be used for
+    site and channel
+    """
+    key = net + separator + sta
+    if len(chan) > 0:
+        key = key + separator + chan
+    if len(loc) > 0:
+        key = key + separator + loc
+    return key
+
+
+class mseed_channel_matcher(NMF):
+    """
+    This class is used to match wf_miniseed to the channel collection using
+    the mseed standard channel string tags net, sta, chan, and (optionally) loc.
+    It can also be used to normalize data saved in wf_TimeSeries where the mseed tags
+    are often altered by MsPASS to change fields like "net" to "READONLYERROR_net".
+    There is an automatic fallback for each of the tags where if the proper
+    name is not found we alway try to use the READONLYERROR_ version before
+    giving up.
+
+    An issue with this matcher is that it is very common to have redundant
+    entries in the channel collection for the same channel of data.  That
+    can happen for a variety of reasons that are harmless.  When that happens
+    the method of this object will normally post an elog message warning of the
+    potential issue.  Those warnings can be silenced by setting verbose
+    in the constructor to False.
+
+    The class also has a cache option that can dramatically improve
+    performance for large data sets.  When using the database option
+    (caching turned off) the normalize method issues a database query
+    at each call.  If applied to a data set with a large number of
+    waveforms that can add up.  We have found the cache algorithm is
+    an order of magnitude or more faster than the database algorithm
+    for typical channel collections assembled from FDSN web services.
+    It is recommended unless the memory foot print is excessive.
+    That too can usually be avoided by using a query to weed out unnecessary
+    channel documents or by editing the channel document to reduce the
+    debris from extraneous data.
+    """
+
+    def __init__(
+        self,
+        db,
+        attributes_to_load=None,
```

Comment:
We need to set the default value of a list/dict argument to None, and initialize it in the function body. That's because the weired feature of python: https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument. 
And I totally agree that we should clarify it in the docstring, I will do that. Thanks for pointing that out.